feat(sqlite): Introduce `RuntimeConfig` which includes `cache_size`

Hywan opened the pull request:
This patch updates `StoreOpenConfig` to hold a new type: `RuntimeConfig`.

This `RuntimeConfig` type is passed to a new `SqliteAsyncConnExt`
method, named `apply_runtime_config`. Depending on the values passed
here, the `optimize`, `cache_size` (new!) and `journal_size_limit`
methods will be called automatically.

The goal of this type is to automate a flow we keep repeating in
all the stores. This is error-prone. This type brings uniformity and
consistency.

This patch also makes all `open_with_pool` methods on the stores private
(they were public before):

1. they were never used as far as I know because getting a `SqlitePool`
   isn't possible since the `pool` attribute is private…
2. it's better to keep control of this flow.

---

* Address https://github.com/matrix-org/matrix-rust-sdk/issues/4801 by introducing `cache_size`
* Built on top of https://github.com/matrix-org/matrix-rust-sdk/pull/4826. The latter must be merged before this one.

poljar commented on crates/matrix-sdk-sqlite/src/utils.rs:
```diff
@@ -105,35 +105,71 @@ pub(crate) trait SqliteAsyncConnExt {
+
+    /// Define the maximum number of pages the SQLite cache can use.
+    ///
+    /// See [`PRAGMA cache_size`] to learn more. This value corresponds to a
+    /// negative `N` in `PRAGMA cache_size = N`.
+    ///
+    /// [`PRAGMA cache_size`]: https://www.sqlite.org/pragma.html#pragma_cache_size
```
Same confusing comment as above.

Hywan commented on crates/matrix-sdk-sqlite/src/utils.rs:
```diff
@@ -105,35 +105,71 @@ pub(crate) trait SqliteAsyncConnExt {
+    ///
+    /// [`PRAGMA cache_size`]: https://www.sqlite.org/pragma.html#pragma_cache_size
+    async fn cache_size(&self, n: u16) -> Result<()> {
+        self.execute_batch(format!("PRAGMA cache_size = -{n};")).await?;
```
If use the value `n` as a negative value here.

Hywan commented on crates/matrix-sdk-sqlite/src/lib.rs:
```diff
@@ -76,21 +79,114 @@ impl SqliteStoreConfig {
+
+    /// Define the maximum number of pages the SQLite cache can use.
+    ///
+    /// See [`PRAGMA cache_size`] to learn more. This value corresponds to a
+    /// negative `N` in `PRAGMA cache_size = N`.
```
We use this value as a negative value here https://github.com/matrix-org/matrix-rust-sdk/pull/4828#discussion_r2016098409. Its semantics is -N.

Hywan replied:
However yes, there is a confusion here. The code is wrong. It represents the number of pages. I'll fix that, it must be a positive N.

poljar replied:
Oh, indeed. Could you then clarify what this means, we don't really need to know about the negation, just that this will correspond to the amount of memory that will be used.

Hywan replied:
Good idea.

Now a question: do we want to configure the cache size as a number of pages or as an absolute value?

Arguments in favour to the latter:

1. It is easier to understand for most people (you would be surprised to learn how many people don't know what a memory page is),
2. A memory page size isn't the same on all systems, so it's harder to control the cache size; it can even change with a system update,
3. The journal size and other sizes in the SDK are expressed in bytes (or kibibytes), not in page size. Consistency matters.

Thought?

poljar replied:
I would say that configuring this in bytes or kibibytes is the simpler API and I don't see the need to exactly control page sizes.

Hywan commented:
Rebased since a small set of patches. I've clarified the confusion around the cache size, and I've added tests for `PRAGMA cache_size` (to check (i) it is passed a negative value, (ii) it is converted from bytes to kibibytes) and `PRAGMA journal_size_limit`.
